File: terraform/lambdas/carts/add_to_cart.py
import json
import boto3
import os
from decimal import Decimal 

dynamodb = boto3.resource('dynamodb')
carts_table = dynamodb.Table(os.environ['CARTS_TABLE'])
products_table = dynamodb.Table(os.environ['PRODUCTS_TABLE'])

def lambda_handler(event, context):
    try:
        body = json.loads(event['body'])
        user_id = body.get('user_id')
        product_id = body.get('product_id')
        quantity = body.get('quantity', 1)
        
        # Verify product exists
        product_response = products_table.get_item(Key={'productId': product_id})
        if 'Item' not in product_response:
            return {
                'statusCode': 404,
                'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'},
                'body': json.dumps({'error': 'Product not found'})
            }
        
        product = product_response['Item']
        
        # Convert price to Decimal for DynamoDB
        price = Decimal(str(product['price'])) if 'price' in product else Decimal('0')
        
        # Add to cart
        carts_table.update_item(
            Key={'user_id': user_id},
            UpdateExpression='SET #items = list_append(if_not_exists(#items, :empty_list), :new_item)',
            ExpressionAttributeNames={'#items': 'items'},
            ExpressionAttributeValues={
                ':empty_list': [],
                ':new_item': [{
                    'product_id': product_id, 
                    'quantity': quantity,
                    'name': product.get('name'),
                    'price': price  # ← Use Decimal instead of float
                }]
            }
        )
        
        return {
            'statusCode': 200,
            'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'},
            'body': json.dumps({'message': 'Item added to cart'})
        }
        
    except Exception as e:
        return {
            'statusCode': 500,
            'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'},
            'body': json.dumps({'error': str(e)})
        }

File: serverless-ecommerce/src/cart/add-to-cart.js
const AWS = require("aws-sdk");
const dynamodb = new AWS.DynamoDB.DocumentClient();

exports.handler = async (event) => {
  try {
    const { userId, productId, quantity } = JSON.parse(event.body);

    // Validate input
    if (!userId || !productId || !quantity) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          error: "Missing required fields: userId, productId, quantity",
        }),
      };
    }

    // Get product to validate existence and price
    const productResult = await dynamodb
      .get({
        TableName: process.env.PRODUCTS_TABLE,
        Key: { productId },
      })
      .promise();

    if (!productResult.Item) {
      return {
        statusCode: 404,
        body: JSON.stringify({ error: "Product not found" }),
      };
    }

    const product = productResult.Item;

    // Check stock availability
    if (product.stock < quantity) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: "Insufficient stock" }),
      };
    }

    // Calculate item total
    const itemTotal = product.price * quantity;

    // Update cart (add or update item)
    const timestamp = new Date().toISOString();
    const ttl = Math.floor(Date.now() / 1000) + 24 * 60 * 60; // 24 hours from now

    await dynamodb
      .update({
        TableName: process.env.CARTS_TABLE,
        Key: { userId },
        UpdateExpression: `SET 
        #items.#productId = :item,
        updatedAt = :updatedAt,
        #ttl = :ttl
        ADD total :itemTotal`,
        ExpressionAttributeNames: {
          "#items": "items",
          "#productId": productId,
          "#ttl": "ttl",
        },
        ExpressionAttributeValues: {
          ":item": {
            productId,
            name: product.name,
            price: product.price,
            quantity,
            itemTotal,
          },
          ":updatedAt": timestamp,
          ":ttl": ttl,
          ":itemTotal": itemTotal,
        },
        ReturnValues: "ALL_NEW",
      })
      .promise();

    return {
      statusCode: 200,
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        message: "Item added to cart",
        product: {
          productId,
          name: product.name,
          price: product.price,
          quantity,
        },
      }),
    };
  } catch (error) {
    console.error("Add to cart error:", error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: "Internal server error" }),
    };
  }
};
